WordPress/gutenberg
# Block Library: Implement Template Part block editing 2.

Alternative to #18925
Follows #18736

## Revisions

- https://github.com/WordPress/gutenberg/pull/19203#issuecomment-567043557
- https://github.com/WordPress/gutenberg/pull/19203#issuecomment-567201464

## Description

This PR continues the block template part work by enabling the editing of existing block template part files as blocks in block templates.

It runs some of the block template override logic in the editor to find the relevant post and create auto drafts for the template parts in it. This is done in the editor settings hook, to be able to send back context to the editor, like the current template's ID, later on, when implementing editing modes.

The edit implementation of the Template Part block can then load the block template part CPT post directly if the block template part has already been customized or it can look for the latest auto draft which was created from the block template part files.

With a way to create and/or load the CPT posts for the template parts, we just needed a way to sync entities with inner blocks. This will be needed for many blocks and the parsing and serialization dance will be the same for all of them so it made sense to introduce ~~2 new hooks~~ a new hook to `core-data` for this:

```js
const [ blocks, onInput, onChange ] = useEntityBlockEditor( kind, type, options );
```

For somewhat low-level manipulation.

~~useEntitySyncedInnerBlocks( kind, type, options )~~

~~For just syncing a block's inner blocks the same way the block editor is synced with the editor. This uses the former hook in its implementation and will be the preferred way to handle things in most blocks, like the Template Part block.~~

Additionally, it looks like the `core-data` functionality of evaluating optimized function edits before saving, was never implemented. I think we decided to back off from it until we decided to "really" support them, and never got around to it. This PR adds it, because unlike the editor which has a single canonical way of saving through `savePost`, entity synced inner blocks can be saved in many different ways/UIs, so it's important for this to happen at the framework level. We'll still need to keep the specific `savePost` behavior, because the editor uses a custom serialization function for backwards compatibility with WordPress in certain cases like when there is an empty paragraph block.

## How to test this?

- Set up the test environment described in #18736's testing instructions.
- Create a block template CPT post called "single" with `/block-templates/single.html`'s contents using the code editor.
- Verify that editing the header block template part and saving both the block template and the block template part propagate the changes everywhere the header is used.

## Types of Changes

*New Feature:* The Template Part block now lets you edit the underlying block template part.
*New Feature:* `core-data` now supports a hook for syncing entities with controlled block/inner block content.
*New Feature:* `InnerBlocks` now supports a controlled mode for syncing blocks with other APIs.

## Checklist:
